### PR TITLE
refactor: Future-proof init/upgrade libraries for other SubDAOs and releases (SC-888)

### PR DESCRIPTION
**Pull request by lucas-manuel**
# Summary of Changes
1. Move ForeignControllerInit and MainnetControllerInit into their own files for easier readability/navigation.
2. Update init libraries to use a standard structure for both foreign and mainnet:
    - `initAlmSystem` initializes almProxy, rateLimits and controller together, doing all sanity checks and all configuration.
    - `upgradeController` initializes a controller within an existing ALM system, deprecating an `oldController`.
    - `_initController` is shared by both of these functions, performing the necessary checks and configuration for initializing just the controller.
3. Update init libraries to no longer set rate limits. This is to be done in spells manually as these configurations will become more sophisticated an custom with newer version. The init/upgrade scripts are intended to be base functionality.
4. Removed sanity checks in init/upgrade libs of values that are either hardcoded or queried from other contracts. Only checking constructor params now.
5. `setRateLimit` was moved to the RateLimitHelpers library. The `rateLimits` param is now `address` instead of `IRateLimits` for easier usage.
6. Staging deployment was refactored to accomodate these changes.
7. Deployments and init/upgrades are now tested in separate files. NOTE: DeployAndInit.t.sol will be renamed to InitAndUpgrade.t.sol in a later PR, this filename was kept as is to have a better diff.

# Testing a Local Staging Deployment
All staging tests are failing, but they can be verified against a local deployment. Set up two anvil nodes forking mainnet and base and do the following commands:

```
export MAINNET_RPC_URL=http://127.0.0.1:8545
export BASE_RPC_URL=http://127.0.0.1:8546

cast rpc --rpc-url="$MAINNET_RPC_URL" anvil_impersonateAccount 0xD1668fB5F690C59Ab4B0CAbAd0f8C1617895052B
cast rpc --rpc-url="$MAINNET_RPC_URL" anvil_impersonateAccount 0xd207546C41F1B2b2cb3857f1De735eCff0bb5FCf
cast send 0x6b175474e89094c44da98b954eedeac495271d0f 'transfer(address,uint256)' $ETH_FROM 1000000000000000000000 --unlocked --from 0xD1668fB5F690C59Ab4B0CAbAd0f8C1617895052B
cast send 0xdC035D45d973E3EC169d2276DDab16f1e407384F 'transfer(address,uint256)' $ETH_FROM 1000000000000000000000 --unlocked --from 0xd207546C41F1B2b2cb3857f1De735eCff0bb5FCf

```

Then replace *latest.json files with release-DATE.json and update RELEASE_DATE in the test.

### CODE REVIEW

File: script/staging/FullStagingDeploy.s.sol [hunk 1]
```diff
@@ -248,146 +312,154 @@ contract FullStagingDeploy is Script {
             slope     : 0
```
Comment (hexonaut) [inline script/staging/FullStagingDeploy.s.sol:312]: I would add this to the `RateLimitHelpers` an unlimited rate limit.

File: script/staging/FullStagingDeploy.s.sol [hunk 2]
```diff
@@ -248,146 +312,154 @@ contract FullStagingDeploy is Script {
...
+    function _setBaseControllerRateLimits(address rateLimits) internal {
+        RateLimitData memory rateLimitData18 = RateLimitData({
+            maxAmount : USDC_UNIT_SIZE * 1e12 * 5,
+            slope     : USDC_UNIT_SIZE * 1e12 / 4 hours
+        });
+        RateLimitData memory rateLimitData6 = RateLimitData({
+            maxAmount : USDC_UNIT_SIZE * 5,
+            slope     : USDC_UNIT_SIZE / 4 hours
+        });
+        RateLimitData memory unlimitedRateLimit = RateLimitData({
```
Comment (hexonaut) [inline script/staging/FullStagingDeploy.s.sol:359]: Same here I would add this as a function to be called, `RateLimitHelpers.unlimitedRateLimitData()` or similar.

File: script/staging/FullStagingDeploy.s.sol [hunk 3]
```diff
@@ -248,146 +312,154 @@ contract FullStagingDeploy is Script {
...
+        RateLimitHelpers.setRateLimitData(RateLimitHelpers.makeAssetKey(psmDepositKey,  usdc),  rateLimits, rateLimitData6,     "usdcDepositDataPsm",   6);
+        RateLimitHelpers.setRateLimitData(RateLimitHelpers.makeAssetKey(psmWithdrawKey, usdc),  rateLimits, rateLimitData6,     "usdcWithdrawDataPsm",  6);
+        RateLimitHelpers.setRateLimitData(RateLimitHelpers.makeAssetKey(psmDepositKey,  usds),  rateLimits, rateLimitData18,    "usdsDepositDataPsm",   18);
+        RateLimitHelpers.setRateLimitData(RateLimitHelpers.makeAssetKey(psmWithdrawKey, usds),  rateLimits, unlimitedRateLimit, "usdsWithdrawDataPsm",  18);
+        RateLimitHelpers.setRateLimitData(RateLimitHelpers.makeAssetKey(psmDepositKey,  susds), rateLimits, rateLimitData18,    "susdsDepositDataPsm",  18);
+        RateLimitHelpers.setRateLimitData(RateLimitHelpers.makeAssetKey(psmWithdrawKey, susds), rateLimits, unlimitedRateLimit, "susdsWithdrawDataPsm", 18);
 
-        // Step 4: Export all deployed addresses
+        // CCTP rate limits
+        RateLimitHelpers.setRateLimitData(foreignController.LIMIT_USDC_TO_CCTP(), rateLimits, rateLimitData6, "usdcToCctpData",           6);
```
Comment (hexonaut) [inline script/staging/FullStagingDeploy.s.sol:391]: This should be unlimited.

File: script/staging/FullStagingDeploy.s.sol [hunk 4]
```diff
@@ -248,146 +312,154 @@ contract FullStagingDeploy is Script {
...
+        RateLimitHelpers.setRateLimitData(domainKeyEthereum,                      rateLimits, rateLimitData6, "cctpToEthereumDomainData", 6);
 
-        ScriptTools.exportContract(mainnet.nameDeps, "freezer",    mainnet.config.readAddress(".freezer"));
-        ScriptTools.exportContract(mainnet.nameDeps, "relayer",    mainnet.config.readAddress(".relayer"));
+        // AAVE rate limits
+        RateLimitHelpers.setRateLimitData(RateLimitHelpers.makeAssetKey(aaveDepositKey,  AUSDC_BASE), rateLimits, rateLimitData6, "usdcDepositDataAave",  6);
+        RateLimitHelpers.setRateLimitData(RateLimitHelpers.makeAssetKey(aaveWithdrawKey, AUSDC_BASE), rateLimits, rateLimitData6, "usdcWithdrawDataAave", 6);
```
Comment (hexonaut) [inline script/staging/FullStagingDeploy.s.sol:396]: This should be unlimited.

File: script/staging/FullStagingDeploy.s.sol [hunk 5]
```diff
@@ -248,146 +312,154 @@ contract FullStagingDeploy is Script {
...
 
-        ScriptTools.exportContract(mainnet.name, "almProxy",   instance.almProxy);
-        ScriptTools.exportContract(mainnet.name, "controller", instance.controller);
-        ScriptTools.exportContract(mainnet.name, "rateLimits", instance.rateLimits);
+        RateLimitHelpers.setRateLimitData(RateLimitHelpers.makeAssetKey(vaultDepositKey,  MORPHO_VAULT_USDC_BASE), rateLimits, rateLimitData6, "usdsDepositDataMorpho", 6);
+        RateLimitHelpers.setRateLimitData(RateLimitHelpers.makeAssetKey(vaultWithdrawKey, MORPHO_VAULT_USDC_BASE), rateLimits, rateLimitData6, "usdsDepositDataMorpho", 6);
```
Comment (hexonaut) [inline script/staging/FullStagingDeploy.s.sol:399]: This should be unlimited.

File: script/staging/test/StagingDeployment.t.sol [hunk 6]
```diff
@@ -342,7 +293,7 @@ contract MainnetStagingDeploymentTests is StagingDeploymentTestBase {
 
         _simulateUsdeBurn(usdeAmount);
 
-        assertGe(usdc.balanceOf(address(almProxy)), startingBalance + 10e6);  // Interest earned
+        assertGe(usdc.balanceOf(address(almProxy)), startingBalance + 10e6 - 1);  // Interest earned (roundign)
```
Comment (hexonaut) [inline script/staging/test/StagingDeployment.t.sol:296]: Typo `rounding`

File: src/RateLimitHelpers.sol [hunk 7]
```diff
@@ -11,4 +18,33 @@ library RateLimitHelpers {
...
+        string        memory name,
+        uint256       decimals
+    )
+        internal
+    {
+        // Handle setting an unlimited rate limit
+        if (data.maxAmount == type(uint256).max) {
+            require(
+                data.slope == 0,
+                string(abi.encodePacked("MainnetControllerInit/invalid-rate-limit-", name))
```
Comment (hexonaut) [inline src/RateLimitHelpers.sol:34]: Should be `RateLimitHelpers` for error.

File: src/RateLimitHelpers.sol [hunk 8]
```diff
@@ -11,4 +18,33 @@ library RateLimitHelpers {
...
+            );
+        }
+        else {
+            require(
+                data.maxAmount <= 1e12 * (10 ** decimals),
+                string(abi.encodePacked("MainnetControllerInit/invalid-max-amount-precision-", name))
```
Comment (hexonaut) [inline src/RateLimitHelpers.sol:40]: Should be `RateLimitHelpers` for error.

File: src/RateLimitHelpers.sol [hunk 9]
```diff
@@ -11,4 +18,33 @@ library RateLimitHelpers {
...
+            );
+            require(
+                data.slope <= 1e12 * (10 ** decimals) / 1 hours,
+                string(abi.encodePacked("MainnetControllerInit/invalid-slope-precision-", name))
```
Comment (hexonaut) [inline src/RateLimitHelpers.sol:44]: Should be `RateLimitHelpers` for error.

**Review by hexonaut — CHANGES_REQUESTED** [review]
Refactor looks good. Some minor changes. Will deploy.

File: script/staging/FullStagingDeploy.s.sol [hunk 10]
```diff
@@ -215,25 +223,81 @@ contract FullStagingDeploy is Script {
...
+
+        ScriptTools.exportContract(mainnet.nameDeps, "freezer", mainnet.config.readAddress(".freezer"));
+        ScriptTools.exportContract(mainnet.nameDeps, "relayer", mainnet.config.readAddress(".relayer"));
+
+        ScriptTools.exportContract(mainnet.name, "almProxy",   controllerInst.almProxy);
+        ScriptTools.exportContract(mainnet.name, "controller", controllerInst.controller);
+        ScriptTools.exportContract(mainnet.name, "rateLimits", controllerInst.rateLimits);
+    }
+
+    // TODO: Remove
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:299]: Was left as a mistake, deleted

File: script/staging/FullStagingDeploy.s.sol [hunk 11]
```diff
@@ -215,25 +223,81 @@ contract FullStagingDeploy is Script {
...
+                psm        : psm,
+                daiUsds    : mainnet.config.readAddress(".daiUsds"),
+                cctp       : mainnet.config.readAddress(".cctpTokenMessenger")
+            });
+
+        MainnetControllerInit.MintRecipient[] memory mintRecipients = new MainnetControllerInit.MintRecipient[](1);
+
+        mintRecipients[0] = MainnetControllerInit.MintRecipient({
+            domain        : CCTPForwarder.DOMAIN_ID_CIRCLE_BASE,
+            mintRecipient : bytes32(uint256(uint160(makeAddr("baseAlmProxy"))))
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:267]: Good point, removed

File: src/RateLimitHelpers.sol (hunk 7 above)
Comment (lucas-manuel) [reply src/RateLimitHelpers.sol:34]: Updated

File: src/RateLimitHelpers.sol (hunk 8 above)
Comment (lucas-manuel) [reply src/RateLimitHelpers.sol:40]: Updated

File: src/RateLimitHelpers.sol (hunk 9 above)
Comment (lucas-manuel) [reply src/RateLimitHelpers.sol:44]: Updated

File: script/staging/FullStagingDeploy.s.sol [hunk 12]
```diff
@@ -248,146 +312,154 @@ contract FullStagingDeploy is Script {
...
+        // USDS mint/burn and cross-chain transfer rate limits
+        RateLimitHelpers.setRateLimitData(mainnetController_.LIMIT_USDS_MINT(),    rateLimits, rateLimitData18,    "usdsMintData",         18);
+        RateLimitHelpers.setRateLimitData(mainnetController_.LIMIT_USDS_TO_USDC(), rateLimits, rateLimitData6,     "usdsToUsdcData",       6);
+        RateLimitHelpers.setRateLimitData(mainnetController_.LIMIT_USDC_TO_CCTP(), rateLimits, unlimitedRateLimit, "usdcToCctpData",       6);
+        RateLimitHelpers.setRateLimitData(domainKeyBase,                           rateLimits, rateLimitData6,     "cctpToBaseDomainData", 6);
+
+        // Ethena-specific rate limits
+        RateLimitHelpers.setRateLimitData(mainnetController_.LIMIT_USDE_MINT(),      rateLimits, rateLimitData6,     "usdeMintData",      6);
+        RateLimitHelpers.setRateLimitData(mainnetController_.LIMIT_USDE_BURN(),      rateLimits, rateLimitData18,    "usdeBurnData",      18);
+        RateLimitHelpers.setRateLimitData(mainnetController_.LIMIT_SUSDE_COOLDOWN(), rateLimits, unlimitedRateLimit, "susdeCooldownData", 18);
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:337]: Updated

File: script/staging/FullStagingDeploy.s.sol [hunk 13]
```diff
@@ -248,146 +312,154 @@ contract FullStagingDeploy is Script {
...
+
+        // 4626 and AAVE deposit/withdraw rate limits
+        RateLimitHelpers.setRateLimitData(ausdcDepositKey,  rateLimits, rateLimitData6,  "ausdcDepositData",  6);
+        RateLimitHelpers.setRateLimitData(ausdcWithdrawKey, rateLimits, rateLimitData6,  "ausdcWithdrawData", 6);
+        RateLimitHelpers.setRateLimitData(ausdsDepositKey,  rateLimits, rateLimitData6,  "ausdsDepositData",  18);
+        RateLimitHelpers.setRateLimitData(ausdsWithdrawKey, rateLimits, rateLimitData6,  "ausdsWithdrawData", 18);
+        RateLimitHelpers.setRateLimitData(susdeDepositKey,  rateLimits, rateLimitData18, "susdeDepositData",  18);
+        RateLimitHelpers.setRateLimitData(susdeWithdrawKey, rateLimits, rateLimitData18, "susdeWithdrawData", 18);
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:345]: Removed

File: script/staging/FullStagingDeploy.s.sol [hunk 14]
```diff
@@ -248,146 +312,154 @@ contract FullStagingDeploy is Script {
...
-        // Configure this after Base ALM Proxy is deployed
-        MintRecipient[] memory mintRecipients = new MintRecipient[](0);
+        MainnetController mainnetController_ = MainnetController(mainnetController);
+
+        bytes32 ausdcDepositKey  = RateLimitHelpers.makeAssetKey(mainnetController_.LIMIT_AAVE_DEPOSIT(),   AUSDC);
+        bytes32 ausdcWithdrawKey = RateLimitHelpers.makeAssetKey(mainnetController_.LIMIT_AAVE_WITHDRAW(),  AUSDC);
+        bytes32 ausdsDepositKey  = RateLimitHelpers.makeAssetKey(mainnetController_.LIMIT_AAVE_DEPOSIT(),   AUSDS);
+        bytes32 ausdsWithdrawKey = RateLimitHelpers.makeAssetKey(mainnetController_.LIMIT_AAVE_WITHDRAW(),  AUSDS);
+        bytes32 susdeDepositKey  = RateLimitHelpers.makeAssetKey(mainnetController_.LIMIT_4626_DEPOSIT(),   address(mainnetController_.susde()));
+        bytes32 susdeWithdrawKey = RateLimitHelpers.makeAssetKey(mainnetController_.LIMIT_4626_WITHDRAW(),  address(mainnetController_.susde()));
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:322]: Removed

File: script/staging/FullStagingDeploy.s.sol (hunk 1 above)
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:312]: Just going to call rateLimits directly with this

File: script/staging/FullStagingDeploy.s.sol (hunk 2 above)
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:359]: Just calling directly

File: script/staging/test/StagingDeployment.t.sol (hunk 6 above)
Comment (lucas-manuel) [reply script/staging/test/StagingDeployment.t.sol:296]: Updated

File: script/staging/FullStagingDeploy.s.sol [hunk 15]
```diff
@@ -248,146 +312,154 @@ contract FullStagingDeploy is Script {
...
+        RateLimitHelpers.setRateLimitData(domainKeyBase,                           rateLimits, rateLimitData6,     "cctpToBaseDomainData", 6);
+
+        // Ethena-specific rate limits
+        RateLimitHelpers.setRateLimitData(mainnetController_.LIMIT_USDE_MINT(),      rateLimits, rateLimitData6,     "usdeMintData",      6);
+        RateLimitHelpers.setRateLimitData(mainnetController_.LIMIT_USDE_BURN(),      rateLimits, rateLimitData18,    "usdeBurnData",      18);
+        RateLimitHelpers.setRateLimitData(mainnetController_.LIMIT_SUSDE_COOLDOWN(), rateLimits, unlimitedRateLimit, "susdeCooldownData", 18);
+
+        // 4626 and AAVE deposit/withdraw rate limits
+        RateLimitHelpers.setRateLimitData(ausdcDepositKey,  rateLimits, rateLimitData6,  "ausdcDepositData",  6);
+        RateLimitHelpers.setRateLimitData(ausdcWithdrawKey, rateLimits, rateLimitData6,  "ausdcWithdrawData", 6);
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:341]: Moved to be unlimited

File: script/staging/FullStagingDeploy.s.sol [hunk 16]
```diff
@@ -248,146 +312,154 @@ contract FullStagingDeploy is Script {
...
+        RateLimitHelpers.setRateLimitData(ausdsDepositKey,  rateLimits, rateLimitData6,  "ausdsDepositData",  18);
+        RateLimitHelpers.setRateLimitData(ausdsWithdrawKey, rateLimits, rateLimitData6,  "ausdsWithdrawData", 18);
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:343]: Moved to be unlimited

File: script/staging/FullStagingDeploy.s.sol (hunk 3 above)
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:391]: Moved to unlimited

File: script/staging/FullStagingDeploy.s.sol (hunk 4 above)
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:396]: Moved to unlimited

File: script/staging/FullStagingDeploy.s.sol (hunk 5 above)
Comment (lucas-manuel) [reply script/staging/FullStagingDeploy.s.sol:399]: Moved to unlimited

**Review by hexonaut — COMMENTED** [review]
All changes look good. Just waiting on moving the unlimited back to their original groupings as discussed in Slack.

**Review by hexonaut — APPROVED** [review]

**Pull request merged**